tools/gap2seq.py

```diff
@@ -0,0 +1,110 @@
+'''
+    Gap2Seq - assembly gap closing tool
+'''
+
+import logging
+import os
+import os.path
+import subprocess
+import shlex
+import shutil
+import tempfile
+import time
+
+import Bio.Seq
+
+import tools
+import tools.samtools
+import util.file
+import util.misc
+
+TOOL_NAME = 'gap2seq'
+TOOL_VERSION = '2.1'
+
+log = logging.getLogger(__name__)
+
+class Gap2SeqTool(tools.Tool):
+    """Tool wrapper for the Gap2Seq gap-closing tool."""
+
+    def __init__(self, install_methods=None):
+        if install_methods is None:
+            install_methods = [tools.CondaPackage(TOOL_NAME, version=TOOL_VERSION, executable='Gap2Seq.sh')]
+        tools.Tool.__init__(self, install_methods=install_methods)
+
+    def version(self):
+        return TOOL_VERSION
+
+    def execute(self, args):    # pylint: disable=W0221
+        tool_cmd = [self.install_and_get_path()] + args
+        log.debug('running gap2seq: ' + ' '.join(tool_cmd))
+        subprocess.check_call(tool_cmd)
+
+    def gapfill(self, in_scaffold, inBam, out_scaffold, solid_kmer_thresholds=(3,2), kmer_sizes=(90, 80, 70, 60, 50, 40, 31),
+                min_gap_to_close=4, gap2seq_opts='', mem_limit_gb=4, threads=1, time_limit_minutes=60):
+        """Try to fill the gaps in the given scaffold, using the reads.
+
+        Inputs:
+            in_scaffold: a FASTA file containing the scaffold.  Each FASTA record corresponds to one
+                segment (for multi-segment genomes).  Contigs within each segment are
+                separated by Ns.  The exact number of Ns between contigs does not matter, as the length of the gap is one 
+                of the things determined by the gap-filling tool.  (But see `min_gap_to_close`).
+            inBam: reads to use for filling the gaps.  Only paired-end reads from the bam file are used, any unpaired reads
+                are ignored.
+           
+        Outputs:
+            out_scaffold: the input scaffold, with some of the gaps between contigs possibly filled.
+
+        Params:
+            solid_kmer_thresholds: kmers must appear at least this many times in the reads to be considered solid.
+                We try gapfilling for all combinations of values of solid_kmer_thresholds and kmer_sizes.
+            kmer_sizes: kmer sizes to use.  We try gapfilling for all combinations of values of solid_kmer_thresholds and kmer_sizes.
+            min_gap_to_close: stop gap-closing if all gaps are no longer than this many Ns
+            gap2seq_opts: extra command-line flags to pass to Gap2Seq
+            mem_limit_gb: max memory to use, in gigabytes
+            threads: number of threads to use
+            time_limit_minutes: stop trying to close more gaps after this many minutes (currently this is a soft/advisory limit)
+        
+        """
+        from os.path import abspath, join
```

Comment:
In general all imports should be at the top for pep8 as well as tooling. If you really want to abbreviate the code I would suggest `join = os.path.join` at the beginning of the function instead.